File: programers_Lv2/피로도.js
function solution(k, dungeons) {
  var answer = 0;

  function permutation(arr) {
    if (arr.length === 0) {
      return [[]];
    }
    const result = [];
    for (let i = 0; i < arr.length; i++) {
      const current = arr[i];
      const remaining = [...arr.slice(0, i), ...arr.slice(i + 1)];
      const subPermutations = permutation(remaining);
      for (let j = 0; j < subPermutations.length; j++) {
        const permutation = [current, ...subPermutations[j]];
        result.push(permutation);
      }
    }
    return result;
  }

  const nPr = permutation(dungeons);
  nPr.forEach((v) => {
    let curK = k;
    let cnt = 0;
    v.forEach((w) => {
      if (curK >= w[0]) {
        curK -= w[1];
        cnt++;
      }
    });
    if (cnt > answer) {
      answer = cnt;
    }
  });

  return answer;
}

console.log(
  solution(80, [
    [80, 20],
    [50, 40],
    [30, 10],
  ])
); // 3

File: programers_Lv2/피로도.py
# https://school.programmers.co.kr/learn/courses/30/lessons/87946
import itertools
def solution(k, dungeons):
    answer = 0
    nPr = itertools.permutations(dungeons, len(dungeons))
    for i in nPr:
        cur_k = k
        cnt = 0
        for j in i:
            if cur_k >= j[0]:
                cur_k -= j[1]
                cnt += 1
        if cnt > answer:
            answer = cnt
    return answer

print(solution(80, [[80,20],[50,40],[30,10]])) # 3
